### ISSUE TITLE
Miscategorization of service labels

### ISSUE THREAD

**Issue by codefromthecrypt** [openzipkin/zipkin#2274]
The current code in the UI prefers the remote service name to the local service name. This is incorrect as the local service name should always be preferred. For example, a client can mislabel the remote side.

That said, there are cases where we want to read the remote service names. For example, root spans who were originated by uninstrumented clients (such as kinesis), or leaf client spans which are calling uninstrumented services (such as s3 or another SaaS). Dependency linker does this, for example. It will read the remote service names for the root and leaves.

One way we can address this is to prefer the remote name only in these use cases. Another way is to display both in these use cases. In the current UI, I think we can only choose to prefer one. In the next UI (project lens), it may be possible to present both. I think haystack-ui presents both for all cases.

**Comment by codefromthecrypt** [comment]
cc @openzipkin/ui

**Comment by codefromthecrypt** [comment]
relates to #1794 

so I think what we can do is float the topic of prefer local vs remote until the rest of v2 is done, preferring local for now. 

I think best case is that the mustache template has a remoteServiceName iff it is a root or leaf
then the UI logic can do whatever it wants. This sort of decoration will be easier when v2 is done.

**Comment by drolando** [comment]
Can you elaborate a bit more on what you mean by "remote service name" and "local service name"?

There are 3 different service names (maybe 4) in a RPC call:
1) CLIENT local service name
2) CLIENT remote service name
3) SERVER local service name

Which one would you show in the UI? I think the one that makes the most sense is 3, using 2 in case there's no server span. Which I believe it's the current UI behavior.

**Comment by codefromthecrypt** [comment]
3 is best. it reflects the same logic as dependency graph. we still have an
issue to help identify remote uninstrumented callers or receivers but I
think we should do that without removing the ability to label the server
correctly.

**Comment by codefromthecrypt** [comment]
fyi @fyi-coursera opened #2345 just now choosing to display only the remote service name when a client-only trace (which would be very weird in practice unless a bug).

I think we had discussions in the past about potentially showing two labels similar to I think how haystack works if I recall. I'm not sure if we want to special-case single-span traces in any way. 

However, I do think that the bottom of the tree (leaves which are clients) should have handling.

**Comment by codefromthecrypt** [comment]
copying discussion from  #2345 as it is very hard to follow decisions and rationale when discussion occurs with partial context in multiple issues.

from @fyi-coursera
> This PR makes the UI use the client remote service name instead of the client local service name if no server local service name exists.

from @drolando 
> I'm +1 on showing the remoteEndpoint name if the server span is missing. For exactly the same reason @fyi-coursera mentioned.
>
> My mysql spans now show up as "serviceX" rather than "mysql" which is very annoying. Same for calls to outside services i.e. AWS or google.
>
> I remember we discussed this a month or so ago and I thought this was already what we agreed upon. And was about to open an issue as well.

from me
>  I think this should only apply to leaf spans. otherwise it will definitely mess up association with the dependency graph.

**Comment by codefromthecrypt** [comment]
As few maintain the UI code I insist people consider the context of this when making suggestions or raising change.

* As noted in this issue, the context for this is either a root span or a leaf client span. In-between will totally mess things up to switch prioritization rules.
* Accordingly, any suggestions must apply IFF the above is the case and not when it is not the case. In other words, tests must prove that code doesn't mess up the intermediate data.
* Not everyone has busted instrumentation. For example, I notice at py_zipkin examples, there is routine mistaken service labels where client and server names are flipped. We should be very careful to get diverse feedback especially if going to change high maintenance code before deferring towards something that may be a side effect of bad instrumentation.


Finally, the prioritization is only needed when there is only one label. We have no need to constrain the lens code to single label.

So, my plea is that when folks suggest things, please consider the whole context above so that we can make a good change and not add new tech debt. Happy to help change, just we need to do it with rigor that means tests and with feedback on both good and busted instrumentation.

cc also @openzipkin/ui

**Comment by codefromthecrypt** [comment]
PS technically checking to see if a span is a leaf implies looking at "children" to see if it is empty
https://github.com/openzipkin/zipkin/blob/master/zipkin-ui/js/component_data/spanNode.js#L31

there is a similar type in zipkin-lens we haven't merged this code to same library yet
